Introduce CUDA-only `_scaled_mm` op

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107178
* __->__ #106844
* #106977


According to https://docs.nvidia.com/cuda/cublas/#id99 only FP8 matrix types can be scaled, and `Float8_e4m3`x`Float8_e4m3` results can be returned as `Float8_e4m3` type, or upcast to `Half`, `BFloat16` or `Float`, but in that case `result_scale` will have no effect as well as `amax` would not be computed.
Optional `bias` argument can also be passed to a function, which should be a vector of either `Half` or `BFloat16`, whose values are added to each row of the result matrix.

See table below for supported input and output types:
| Mat1 type  | Mat2 type | Bias type | Output types |
| ----------- | ----------- | ----------- | ----------- |
| Float8_e4m3  | Float8_e4m3  | Float16  | Float8_e4m3, Float16 |
| Float8_e4m3  | Float8_e4m3  | BFloat16 | Float8_e4m3, BFloat16, Float |
| Float8_e5m2  | Float8_e4m3  | Float16 |  Float8_e4m3, Float8_e5m2, Float16  |
| Float8_e5m2  | Float8_e4m3  | BFloat16 |  Float8_e4m3, Float8_e5m2, BFloat16, Float |
| Float8_e4m3  | Float8_e5m2  | Float16 |  Float8_e4m3, Float8_e5m2, Float16 |
| Float8_e4m3  | Float8_e5m2  | BFloat16 |  Float8_e4m3, Float8_e5m2,  BFloat16, Float |
| Float8_e4m3  | Float8_e5m2  | Not supported | Not supported |

Skip decomposition implementation until fp8-on-triton story is better defined, Potential decomposition can look something like the following:
```python
@register_decomposition(aten._scaled_mm)
def _scaled_mm(
    mat1: Tensor,
    mat2: Tensor,
    *,
    dtype: Optional[torch.dtype] = None,
    scale_a: Optional[Tensor] = None,
    scale_b: Optional[Tensor] = None,
    scale_result: Optional[Tensor] = None,
) -> Tuple[Tensor, Tensor]:
    rc = torch.mm(mat1.to(torch.float32), mat2.to(torch.float32))
    rc = scale_a * rc if scale_a is not None else rc
    rc = scale_b * rc if scale_b is not None else rc
    rc = scale_result * rc if scale_result is not None else rc
    rc = rc.to(dtype if dtype is not None else mat1.dtype)
    return rc, torch.tensor(0.0, device=mat1.device)
```


Known limitations:
  - Only works for matrix sizes divisible by 16
  - 1st operand must be in row-major and 2nd in column-major orders (i.e. if `x` and `y` are contiguous, than only `torch._scaled_mm(x, y.t())` will work)

malfet committed:
Introduce CUDA-only `_scaled_mm` op

[ghstack-poisoned]

malfet committed:
Update on "Introduce CUDA-only `_scaled_mm` op"




[ghstack-poisoned]

vkuzo commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -876,6 +880,111 @@ template void gemm_and_bias(
+  CuBlasLtMatmulDescriptor computeDesc(computeType, scaleType);
+  computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_TRANSA, _cublasOpFromChar(transa));
+  computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_TRANSB, _cublasOpFromChar(transb));
+  if (isFloat8Type(mat1_dtype)) {
```
does this also need to check if `mat2_dtype` is a float8 variant?

malfet replied:
Not really, as for CuBLAS will currently error out if one matrix is fp8 and another is not, that should be checked at the higher level (because this one will raise `cublasLtMatmulAlgoGetHeuristic` error)

malfet committed:
Update on "Introduce CUDA-only `_scaled_mm` op"




[ghstack-poisoned]

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6642,6 +6642,16 @@
   structured_delegate: _addmm_activation.out
   variants: function, method
 
+- func: _scaled_mm(Tensor self, Tensor mat2, *, ScalarType? dtype=None, Tensor? scale_a=None, Tensor? scale_b=None, Tensor? scale_result=None) -> (Tensor, Tensor)
```
What is the dtype argument doing here? If it just for output dtype: `out_dtype` would be a better name>.

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6642,6 +6642,16 @@
   structured_delegate: _addmm_activation.out
   variants: function, method
 
+- func: _scaled_mm(Tensor self, Tensor mat2, *, ScalarType? dtype=None, Tensor? scale_a=None, Tensor? scale_b=None, Tensor? scale_result=None) -> (Tensor, Tensor)
```
nit: you can name these outputs so that they will be named on the python side when returned automatically!

malfet replied:
I though this is a discouraged behavior (as it requires allowlisting it)

albanD replied:
Yeah that is surprising...  not needed here for sure!

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6642,6 +6642,16 @@
+  dispatch:
+    CUDA: _scaled_mm_cuda
+
+- func: _scaled_mm.out(Tensor self, Tensor mat2, *, ScalarType? dtype=None, Tensor? scale_a=None, Tensor? scale_b=None, Tensor? scale_result=None, Tensor(a!) out, Tensor(b!) out_amax) -> (Tensor(a!), Tensor(b!))
```
I would suggest you do this as a structured kernel (similar to addmm today). You only have to replace this with a meta impl (which is trivial) the right way. This will allow you to remove a bunch of manual cpp code.

malfet replied:
Hmm, are you saying that structured kernels will take PyThon meta implementation?

albanD commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -197,6 +198,9 @@ static size_t _getWorkspaceSize() {
   return workspace_size;
 }
 
+inline bool isFloat8Type(::c10::ScalarType t) {
```
Should we have this with the other dtype utils in c10::ScalarType as I'm sure we'll need this in other places?

malfet replied:
Good point, refactored into https://github.com/pytorch/pytorch/pull/106977

malfet committed:
Update on "Introduce CUDA-only `_scaled_mm` op"




[ghstack-poisoned]

albanD commented on aten/src/ATen/native/cuda/Blas.cpp:
```diff
@@ -713,4 +714,71 @@ Tensor _int_mm_cuda(const Tensor& self, const Tensor& mat2) {
+          Tensor& out, Tensor& amax) {
+  check_scaled_mm(mat1, mat2);
+  TensorArg targs[]{{out, "out", 0}, {amax, "amax", 1}, {mat1, "mat1", 2}, {mat2, "mat2", 3}};
+  checkAllSameGPU(__func__, targs);
```
You need to check scale devices as well?

albanD commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -876,6 +880,111 @@ template void gemm_and_bias(
+  if (isFloat8Type(mat1_dtype)) {
+    computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_A_SCALE_POINTER, mat1_scale_ptr);
+    computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_B_SCALE_POINTER, mat2_scale_ptr);
+    computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_D_SCALE_POINTER, result_scale_ptr);
```
There is no result_scale used if only mat2 is fp8?

malfet replied:
One can not multiple float8 and non-float8 matrices at the moment, so nothing will work

albanD commented on aten/src/ATen/cuda/CUDABlas.cpp:
```diff
@@ -876,6 +880,111 @@ template void gemm_and_bias(
+  CuBlasLtMatrixLayout Cdesc(isFloat8Type(result_dtype) ? CUDA_R_16F : out_type, m, n, result_ld);
+  CuBlasLtMatrixLayout Ddesc(out_type, m, n, result_ld);
+  size_t workspaceSize = 32*1024*1024;
+  auto workspace = at::empty(
```
There is a special function we have to get this. I know structured sparsity PR uses it if you can't find it.

albanD replied:
```c++
    auto& allocator = *::c10::cuda::CUDACachingAllocator::get();
    auto compressedBufferPtr = allocator.allocate(compressed_buffer_size);
```
This is how it is done there

vkuzo commented on torch/_decomp/decompositions.py:
```diff
@@ -3709,6 +3709,19 @@ def multilabel_margin_loss_forward(
 
 
+@register_decomposition(aten._scaled_mm)
+def _scaled_mm(mat1: Tensor, mat2: Tensor, *,
```
just checking, should we have a decomposition for this?  What I see in this decomposition code would be an emulation of what this op is doing, but it's not actually the same thing.  I would have expected "no decomposition exists" unless we actually have primitive ops which can implement this operation exactly.

malfet committed:
Update on "Introduce CUDA-only `_scaled_mm` op"


Skip decomposition implementation until fp8-on-triton path has been cleaned up, but potential decomposition can look something like the following:
```python
register_decomposition(aten._scaled_mm)
def _scaled_mm(
    mat1: Tensor,
    mat2: Tensor,
    *,
    dtype: Optional[torch.dtype] = None,
    scale_a: Optional[Tensor] = None,
    scale_b: Optional[Tensor] = None,
    scale_result: Optional[Tensor] = None,
) -> Tuple[Tensor, Tensor]:
    rc = torch.mm(mat1.to(torch.float32), mat2.to(torch.float32))
    rc = scale_a * rc if scale_a is not None else rc
    rc = scale_b * rc if scale_b is not None else rc
    rc = scale_result * rc if scale_result is not None else rc
    rc = rc.to(dtype if dtype is not None else mat1.dtype)
    return rc, torch.tensor(0.0, device=mat1.device)
```

[ghstack-poisoned]

malfet committed:
Update on "Introduce CUDA-only `_scaled_mm` op"


Skip decomposition implementation until fp8-on-triton path has been cleaned up, but potential decomposition can look something like the following:
```python
register_decomposition(aten._scaled_mm)
def _scaled_mm(
    mat1: Tensor,
    mat2: Tensor,
    *,
    dtype: Optional[torch.dtype] = None,
    scale_a: Optional[Tensor] = None,
    scale_b: Optional[Tensor] = None,
    scale_result: Optional[Tensor] = None,
) -> Tuple[Tensor, Tensor]:
    rc = torch.mm(mat1.to(torch.float32), mat2.to(torch.float32))
    rc = scale_a * rc if scale_a is not None else rc
    rc = scale_b * rc if scale_b is not None else rc
    rc = scale_result * rc if scale_result is not None else rc
    rc = rc.to(dtype if dtype is not None else mat1.dtype)
    return rc, torch.tensor(0.0, device=mat1.device)
```

[ghstack-poisoned]

malfet committed:
Update on "Introduce CUDA-only `_scaled_mm` op"


Skip decomposition implementation until fp8-on-triton path has been cleaned up, but potential decomposition can look something like the following:
```python
register_decomposition(aten._scaled_mm)
def _scaled_mm(
    mat1: Tensor,
    mat2: Tensor,
    *,
    dtype: Optional[torch.dtype] = None,
    scale_a: Optional[Tensor] = None,
    scale_b: Optional[Tensor] = None,
    scale_result: Optional[Tensor] = None,
) -> Tuple[Tensor, Tensor]:
    rc = torch.mm(mat1.to(torch.float32), mat2.to(torch.float32))
    rc = scale_a * rc if scale_a is not None else rc
    rc = scale_b * rc if scale_b is not None else rc
    rc = scale_result * rc if scale_result is not None else rc
    rc = rc.to(dtype if dtype is not None else mat1.dtype)
    return rc, torch.tensor(0.0, device=mat1.device)
```

[ghstack-poisoned]

malfet committed:
Update on "Introduce CUDA-only `_scaled_mm` op"


Skip decomposition implementation until fp8-on-triton path has been cleaned up, but potential decomposition can look something like the following:
```python
register_decomposition(aten._scaled_mm)
def _scaled_mm(
    mat1: Tensor,
    mat2: Tensor,
    *,
    dtype: Optional[torch.dtype] = None,
    scale_a: Optional[Tensor] = None,
    scale_b: Optional[Tensor] = None,
    scale_result: Optional[Tensor] = None,
) -> Tuple[Tensor, Tensor]:
    rc = torch.mm(mat1.to(torch.float32), mat2.to(torch.float32))
    rc = scale_a * rc if scale_a is not None else rc
    rc = scale_b * rc if scale_b is not None else rc
    rc = scale_result * rc if scale_result is not None else rc
    rc = rc.to(dtype if dtype is not None else mat1.dtype)
    return rc, torch.tensor(0.0, device=mat1.device)
```

[ghstack-poisoned]

malfet committed:
Update on "Introduce CUDA-only `_scaled_mm` op"


Skip decomposition implementation until fp8-on-triton path has been cleaned up, but potential decomposition can look something like the following:
```python
register_decomposition(aten._scaled_mm)
def _scaled_mm(
    mat1: Tensor,
    mat2: Tensor,
    *,
    dtype: Optional[torch.dtype] = None,
    scale_a: Optional[Tensor] = None,
    scale_b: Optional[Tensor] = None,
    scale_result: Optional[Tensor] = None,
) -> Tuple[Tensor, Tensor]:
    rc = torch.mm(mat1.to(torch.float32), mat2.to(torch.float32))
    rc = scale_a * rc if scale_a is not None else rc
    rc = scale_b * rc if scale_b is not None else rc
    rc = scale_result * rc if scale_result is not None else rc
    rc = rc.to(dtype if dtype is not None else mat1.dtype)
    return rc, torch.tensor(0.0, device=mat1.device)
```

[ghstack-poisoned]

malfet committed:
Update on "Introduce CUDA-only `_scaled_mm` op"


Skip decomposition implementation until fp8-on-triton path has been cleaned up, but potential decomposition can look something like the following:
```python
register_decomposition(aten._scaled_mm)
def _scaled_mm(
    mat1: Tensor,
    mat2: Tensor,
    *,
    dtype: Optional[torch.dtype] = None,
    scale_a: Optional[Tensor] = None,
    scale_b: Optional[Tensor] = None,
    scale_result: Optional[Tensor] = None,
) -> Tuple[Tensor, Tensor]:
    rc = torch.mm(mat1.to(torch.float32), mat2.to(torch.float32))
    rc = scale_a * rc if scale_a is not None else rc
    rc = scale_b * rc if scale_b is not None else rc
    rc = scale_result * rc if scale_result is not None else rc
    rc = rc.to(dtype if dtype is not None else mat1.dtype)
    return rc, torch.tensor(0.0, device=mat1.device)
```

[ghstack-poisoned]

malfet committed:
Update on "Introduce CUDA-only `_scaled_mm` op"



According to https://docs.nvidia.com/cuda/cublas/#id99 only FP8 matrix types can be scaled, and `Float8_e4m3`x`Float8_e4m3` results can be returned as `Float8_e4m3` type, or upcast to `Half`, `BFloat16` or `Float`, but in that case `result_scale` will have no effect as well as `amax` would not be computed.

See table below for supported input and output types:
| Mat1 Type  | Mat2 Type | Output Type |
| ----------- | ----------- | ----------- |
| Float8_e4m3  | Float8_e4m3  | Float8_e4m3, Float16, BFloat16, Float |
| Float8_e5m2  | Float8_e4m3  | Float8_e4m3, Float8_e5m2, Float16, BFloat16, Float |
| Float8_e4m3  | Float8_e5m2  | Float8_e4m3, Float8_e5m2, Float16, BFloat16, Float |
| Float8_e4m3  | Float8_e5m2  | Not supported |

Skip decomposition implementation until fp8-on-triton story is better defined, Potential decomposition can look something like the following:
```python
register_decomposition(aten._scaled_mm)
def _scaled_mm(
    mat1: Tensor,
    mat2: Tensor,
    *,
    dtype: Optional[torch.dtype] = None,
    scale_a: Optional[Tensor] = None,
    scale_b: Optional[Tensor] = None,
    scale_result: Optional[Tensor] = None,
) -> Tuple[Tensor, Tensor]:
    rc = torch.mm(mat1.to(torch.float32), mat2.to(torch.float32))
    rc = scale_a * rc if scale_a is not None else rc
    rc = scale_b * rc if scale_b is not None else rc
    rc = scale_result * rc if scale_result is not None else rc
    rc = rc.to(dtype if dtype is not None else mat1.dtype)
    return rc, torch.tensor(0.0, device=mat1.device)
```


Known limitations:
  - Only works for matrix sizes divisible by 16
  - 1st operand must be in row-major and 2nd in column-major orders (i.e. if `x` and `y` are contiguous, than only `torch._scaled_mm(x, y.t())` will work)

[ghstack-poisoned]

malfet committed:
Update on "Introduce CUDA-only `_scaled_mm` op"



According to https://docs.nvidia.com/cuda/cublas/#id99 only FP8 matrix types can be scaled, and `Float8_e4m3`x`Float8_e4m3` results can be returned as `Float8_e4m3` type, or upcast to `Half`, `BFloat16` or `Float`, but in that case `result_scale` will have no effect as well as `amax` would not be computed.

See table below for supported input and output types:
| Mat1 Type  | Mat2 Type | Output Type |
| ----------- | ----------- | ----------- |
| Float8_e4m3  | Float8_e4m3  | Float8_e4m3, Float16, BFloat16, Float |
| Float8_e5m2  | Float8_e4m3  | Float8_e4m3, Float8_e5m2, Float16, BFloat16, Float |
| Float8_e4m3  | Float8_e5m2  | Float8_e4m3, Float8_e5m2, Float16, BFloat16, Float |
| Float8_e4m3  | Float8_e5m2  | Not supported |

Skip decomposition implementation until fp8-on-triton story is better defined, Potential decomposition can look something like the following:
```python
register_decomposition(aten._scaled_mm)
def _scaled_mm(
    mat1: Tensor,
    mat2: Tensor,
    *,
    dtype: Optional[torch.dtype] = None,
    scale_a: Optional[Tensor] = None,
    scale_b: Optional[Tensor] = None,
    scale_result: Optional[Tensor] = None,
) -> Tuple[Tensor, Tensor]:
    rc = torch.mm(mat1.to(torch.float32), mat2.to(torch.float32))
    rc = scale_a * rc if scale_a is not None else rc
    rc = scale_b * rc if scale_b is not None else rc
    rc = scale_result * rc if scale_result is not None else rc
    rc = rc.to(dtype if dtype is not None else mat1.dtype)
    return rc, torch.tensor(0.0, device=mat1.device)
```


Known limitations:
  - Only works for matrix sizes divisible by 16
  - 1st operand must be in row-major and 2nd in column-major orders (i.e. if `x` and `y` are contiguous, than only `torch._scaled_mm(x, y.t())` will work)

[ghstack-poisoned]

drisspg commented on aten/src/ATen/native/cuda/Blas.cpp:
```diff
@@ -713,4 +714,77 @@ Tensor _int_mm_cuda(const Tensor& self, const Tensor& mat2) {
+          c10::optional<c10::ScalarType> out_dtype,
+          const c10::optional<at::Tensor>& scale_a,
+          const c10::optional<at::Tensor>& scale_b,
+          const c10::optional<at::Tensor>& scale_result,
```
I think some documentation on what scale_result would also be helpful

vkuzo commented on aten/src/ATen/native/cuda/Blas.cpp:
```diff
@@ -713,4 +714,77 @@ Tensor _int_mm_cuda(const Tensor& self, const Tensor& mat2) {
 
+// Computes matrix multiply while applying scaling to input and output matrices and computes amax
+// Scales are only applicable when matrices are of Float8 type and assumbed to be equal to 1.0 by default
+// If output matrix type is 16 or 32-bit type, neither scale is applied nor amax is computed
```
hmm, if the inputs are float8 and the output is float16/float32, don't we still need scales for the inputs?

malfet replied:
Ok, let me clarify that it only applies to output scale (i.e. input scales do apply, but output scale is no-op
Another thing probably worth documenting, that amax is computed before scaling

drisspg commented on test/test_matmul_cuda.py:
```diff
@@ -175,7 +176,53 @@ def _expand_to_batch(t: torch.Tensor):
+            self._test_tautological_mm(device, out_dtype=torch.float8_e5m2)
+
+    def test_float8_scale(self, device) -> None:
+        size = (16, 16)
```
If I change this to (7x7) I get:
```
RuntimeError: CUDA error: CUBLAS_STATUS_NOT_SUPPORTED when calling `cublasLtMatmulAlgoGetHeuristic( ltHandle, computeDesc.descriptor(), Adesc.descriptor(), Bdesc.descriptor(), Cdesc.descriptor(), Ddesc.descriptor(), preference.descriptor(), 1, &heuristicResult, &returnedResult)`

```

malfet replied:
Yep, see limitations line: `Only works for matrix sizes divisible by 16`

drisspg reviewed (commented):
We might need some alignment checks

albanD reviewed (approved):
More alignment check hardening (and other improvements)will need to be added but the basic structure sounds good!

malfet committed:
Update on "Introduce CUDA-only `_scaled_mm` op"



According to https://docs.nvidia.com/cuda/cublas/#id99 only FP8 matrix types can be scaled, and `Float8_e4m3`x`Float8_e4m3` results can be returned as `Float8_e4m3` type, or upcast to `Half`, `BFloat16` or `Float`, but in that case `result_scale` will have no effect as well as `amax` would not be computed.

See table below for supported input and output types:
| Mat1 Type  | Mat2 Type | Output Type |
| ----------- | ----------- | ----------- |
| Float8_e4m3  | Float8_e4m3  | Float8_e4m3, Float16, BFloat16, Float |
| Float8_e5m2  | Float8_e4m3  | Float8_e4m3, Float8_e5m2, Float16, BFloat16, Float |
| Float8_e4m3  | Float8_e5m2  | Float8_e4m3, Float8_e5m2, Float16, BFloat16, Float |
| Float8_e4m3  | Float8_e5m2  | Not supported |

Skip decomposition implementation until fp8-on-triton story is better defined, Potential decomposition can look something like the following:
```python
register_decomposition(aten._scaled_mm)
def _scaled_mm(
    mat1: Tensor,
    mat2: Tensor,
    *,
    dtype: Optional[torch.dtype] = None,
    scale_a: Optional[Tensor] = None,
    scale_b: Optional[Tensor] = None,
    scale_result: Optional[Tensor] = None,
) -> Tuple[Tensor, Tensor]:
    rc = torch.mm(mat1.to(torch.float32), mat2.to(torch.float32))
    rc = scale_a * rc if scale_a is not None else rc
    rc = scale_b * rc if scale_b is not None else rc
    rc = scale_result * rc if scale_result is not None else rc
    rc = rc.to(dtype if dtype is not None else mat1.dtype)
    return rc, torch.tensor(0.0, device=mat1.device)
```


Known limitations:
  - Only works for matrix sizes divisible by 16
  - 1st operand must be in row-major and 2nd in column-major orders (i.e. if `x` and `y` are contiguous, than only `torch._scaled_mm(x, y.t())` will work)

[ghstack-poisoned]

malfet commented:
@pytorchbot merge

malfet committed:
Update on "Introduce CUDA-only `_scaled_mm` op"



According to https://docs.nvidia.com/cuda/cublas/#id99 only FP8 matrix types can be scaled, and `Float8_e4m3`x`Float8_e4m3` results can be returned as `Float8_e4m3` type, or upcast to `Half`, `BFloat16` or `Float`, but in that case `result_scale` will have no effect as well as `amax` would not be computed.
Optional `bias` argument can also be passed to a function, which should be a vector of either `Half` or `BFloat16`, whose values are added to each row of the result matrix.

See table below for supported input and output types:
| Mat1 type  | Mat2 type | Bias type | Output types |
| ----------- | ----------- | ----------- | ----------- |
| Float8_e4m3  | Float8_e4m3  | Float16  | Float8_e4m3, Float16 |
| Float8_e4m3  | Float8_e4m3  | BFloat16 | Float8_e4m3, BFloat16, Float |
| Float8_e5m2  | Float8_e4m3  | Float16 |  Float8_e4m3, Float8_e5m2, Float16  |
| Float8_e5m2  | Float8_e4m3  | BFloat16 |  Float8_e4m3, Float8_e5m2, BFloat16, Float |
| Float8_e4m3  | Float8_e5m2  | Float16 |  Float8_e4m3, Float8_e5m2, Float16 |
| Float8_e4m3  | Float8_e5m2  | BFloat16 |  Float8_e4m3, Float8_e5m2,  BFloat16, Float |
| Float8_e4m3  | Float8_e5m2  | Not supported | Not supported |

Skip decomposition implementation until fp8-on-triton story is better defined, Potential decomposition can look something like the following:
```python
register_decomposition(aten._scaled_mm)
def _scaled_mm(
    mat1: Tensor,
    mat2: Tensor,
    *,
    dtype: Optional[torch.dtype] = None,
    scale_a: Optional[Tensor] = None,
    scale_b: Optional[Tensor] = None,
    scale_result: Optional[Tensor] = None,
) -> Tuple[Tensor, Tensor]:
    rc = torch.mm(mat1.to(torch.float32), mat2.to(torch.float32))
    rc = scale_a * rc if scale_a is not None else rc
    rc = scale_b * rc if scale_b is not None else rc
    rc = scale_result * rc if scale_result is not None else rc
    rc = rc.to(dtype if dtype is not None else mat1.dtype)
    return rc, torch.tensor(0.0, device=mat1.device)
```


Known limitations:
  - Only works for matrix sizes divisible by 16
  - 1st operand must be in row-major and 2nd in column-major orders (i.e. if `x` and `y` are contiguous, than only `torch._scaled_mm(x, y.t())` will work)

[ghstack-poisoned]

malfet commented:
@pytorchbot merge

malfet commented:
@pytorchbot merge -i

izaitsevfb commented:
@pytorchbot revert -m 'Breaks internal builds' -c ghfirst

Reverting in the interest of unblocking diff train triaging.

Failure context:
```
aten/src/ATen/cuda/CUDABlas.cpp:907:28: error: use of undeclared identifier 'CUBLASLT_MATMUL_DESC_D_SCALE_POINTER'; did you mean 'CUBLASLT_MATMUL_DESC_BIAS_POINTER'?
computeDesc.setAttribute(CUBLASLT_MATMUL_DESC_D_SCALE_POINTER, result_scale_ptr);
```
Please see [D48377631](https://www.internalfb.com/diff/D48377631)
